presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java

```diff
@@ -19,6 +19,7 @@
 import com.google.common.util.concurrent.ListenableFuture;
```

Comment:
Not controversial either. But do the changes around managing the futre have an impact? You are removing some volatile reads yes, but you are also adding conditional logic under the lock.

Reply:
True, but this will be a "shallow check" in that we're just comparing the reference value and not accessing the target of the reference. As such, one would expect this to amount to a very-local CPU cache entry because we've already taken ownership of the object monitor.

Ultimately it's not obvious whether to unconditionally save and clear the listener without branching is better here. It could well be that the CPU wouldn't issue a store operation if the value was null and null was being set.